# Assignment 1. House Pricing.
by Addisu Amare

---

By the end of this task you will be able to manipulate huge tabular data:
1. Compute different column's statistics (min, max, mean, quantiles etc.);
2. Select observations/features by condition/index;
3. Create new non-linear combinations of the columns (feature engineering);
4. Perform automated data cleaning;

and more.

---

For those who are not familiar with `pandas` we recommend these (alternative) tutorials:

1. Single notebook, covers basic pandas functionality (starting with renaming columns ending with using map, apply etc) ~ 30 short examples with links on videos https://nbviewer.jupyter.org/github/justmarkham/pandas-videos/blob/master/pandas.ipynb . Highly recommended for everyone. (about 1-3 hours to go through)

2. https://github.com/guipsamora/pandas_exercises/ 11 topics covering all essential functionality with excersises (with solutions).

This task will be an easy ride after these tutorials.

---

We are using house sale price data from King County, Wahington, USA. This dataset is in public domain and can be obtained from Kaggle: https://www.kaggle.com/harlfoxem/housesalesprediction

You need to place `kc_house_data.csv` file in the same directory as this notebook.

In [ ]:
import pandas as pd
import numpy as np
import random

random.seed(42)
np.random.seed(42)

In [ ]:
from google.colab import files
uploaded=files.upload()

Saving kc_house_data.csv to kc_house_data (13).csv


# Load the data

In [ ]:
df=pd.read_csv("kc_house_data.csv")

df

id             date     price  bedrooms  bathrooms  \
0      7129300520  20141013T000000  221900.0         3       1.00   
1      6414100192  20141209T000000  538000.0         3       2.25   
2      5631500400  20150225T000000  180000.0         2       1.00   
3      2487200875  20141209T000000  604000.0         4       3.00   
4      1954400510  20150218T000000  510000.0         3       2.00   
...           ...              ...       ...       ...        ...   
21608   263000018  20140521T000000  360000.0         3       2.50   
21609  6600060120  20150223T000000  400000.0         4       2.50   
21610  1523300141  20140623T000000  402101.0         2       0.75   
21611   291310100  20150116T000000  400000.0         3       2.50   
21612  1523300157  20141015T000000  325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  condition  grade  \
0             1180      5650     1.0           0     0          3      7   
1             2570      7242     2.0           0     0          3      7   
2              770     10000     1.0           0     0          3      6   
3             1960      5000     1.0           0     0          5      7   
4             1680      8080     1.0           0     0          3      8   
...            ...       ...     ...         ...   ...        ...    ...   
21608         1530      1131     3.0           0     0          3      8   
21609         2310      5813     2.0           0     0          3      8   
21610         1020      1350     2.0           0     0          3      7   
21611         1600      2388     2.0           0     0          3      8   
21612         1020      1076     2.0           0     0          3      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
0            1180              0      1955             0    98178  47.5112   
1            2170            400      1951          1991    98125  47.7210   
2             770              0      1933             0    98028  47.7379   
3            1050            910      1965             0    98136  47.5208   
4            1680              0      1987             0    98074  47.6168   
...           ...            ...       ...           ...      ...      ...   
21608        1530              0      2009             0    98103  47.6993   
21609        2310              0      2014             0    98146  47.5107   
21610        1020              0      2009             0    98144  47.5944   
21611        1600              0      2004             0    98027  47.5345   
21612        1020              0      2008             0    98144  47.5941   

          long  sqft_living15  sqft_lot15  
0     -122.257           1340        5650  
1     -122.319           1690        7639  
2     -122.233           2720        8062  
3     -122.393           1360        5000  
4     -122.045           1800        7503  
...        ...            ...         ...  
21608 -122.346           1530        1509  
21609 -122.362           1830        7200  
21610 -122.299           1020        2007  
21611 -122.069           1410        1287  
21612 -122.299           1020        1357  

[21613 rows x 21 columns]

# Observe top 10 observations (int)

# Q1.1 What is the price of a house with `id` == 7237550310?
# Q1.2 How many bedrooms has a house with `id` == 7237550310?
# Q1.3 When was the house with `id` ==2414600126  built (`yr_built`)?
# Q1.4 What is the `grade` of a house with `id` == 5631500400?
# Q1.5 When was the house with `id` == 6414100192 renovated (`yr_renovated`)?


In [ ]:
df.head(10)

In [ ]:
#Q1.2
df.loc[df.id==7237550310].bedrooms.values[0]

4

In [ ]:
#Answer for Q1.3
df.loc[df.id==2414600126].yr_built.values[0]

1960

In [ ]:
# Observe last 10 observations (int)

# Q2.1 What is the price of a house with `id` == 263000018?
# Q2.2 How many bedrooms has a house with `id` == 291310100?
# Q2.3 When was the house with `id` == 1523300141 built (`yr_built`)?
# Q2.4 How many floors house with `id` == 2997800021 has?
# Q2.5 What is the zipcode of the house with `id` == 7852140040?


In [ ]:
df.tail(10)

id             date      price  bedrooms  bathrooms  \
21603  7852140040  20140825T000000   507250.0         3       2.50   
21604  9834201367  20150126T000000   429000.0         3       2.00   
21605  3448900210  20141014T000000   610685.0         4       2.50   
21606  7936000429  20150326T000000  1007500.0         4       3.50   
21607  2997800021  20150219T000000   475000.0         3       2.50   
21608   263000018  20140521T000000   360000.0         3       2.50   
21609  6600060120  20150223T000000   400000.0         4       2.50   
21610  1523300141  20140623T000000   402101.0         2       0.75   
21611   291310100  20150116T000000   400000.0         3       2.50   
21612  1523300157  20141015T000000   325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  condition  grade  \
21603         2270      5536     2.0           0     0          3      8   
21604         1490      1126     3.0           0     0          3      8   
21605         2520      6023     2.0           0     0          3      9   
21606         3510      7200     2.0           0     0          3      9   
21607         1310      1294     2.0           0     0          3      8   
21608         1530      1131     3.0           0     0          3      8   
21609         2310      5813     2.0           0     0          3      8   
21610         1020      1350     2.0           0     0          3      7   
21611         1600      2388     2.0           0     0          3      8   
21612         1020      1076     2.0           0     0          3      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
21603        2270              0      2003             0    98065  47.5389   
21604        1490              0      2014             0    98144  47.5699   
21605        2520              0      2014             0    98056  47.5137   
21606        2600            910      2009             0    98136  47.5537   
21607        1180            130      2008             0    98116  47.5773   
21608        1530              0      2009             0    98103  47.6993   
21609        2310              0      2014             0    98146  47.5107   
21610        1020              0      2009             0    98144  47.5944   
21611        1600              0      2004             0    98027  47.5345   
21612        1020              0      2008             0    98144  47.5941   

          long  sqft_living15  sqft_lot15  
21603 -121.881           2270        5731  
21604 -122.288           1400        1230  
21605 -122.167           2520        6023  
21606 -122.398           2050        6200  
21607 -122.409           1330        1265  
21608 -122.346           1530        1509  
21609 -122.362           1830        7200  
21610 -122.299           1020        2007  
21611 -122.069           1410        1287  
21612 -122.299           1020        1357

In [ ]:
#Q2.4
int(df.loc[df.id== 2997800021].floors.values[0])

2

In [ ]:
#Q2.5
df.loc[df.id== 7852140040].zipcode.values[0]

98065

In [ ]:
# Increase maximal displayed columns

x=pd.set_option('display.max_columns', 30)

In [ ]:
# Observe top 10 observations again
# Is there any new columns displayed?
df.head(10)

id             date      price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000   221900.0         3       1.00         1180   
1  6414100192  20141209T000000   538000.0         3       2.25         2570   
2  5631500400  20150225T000000   180000.0         2       1.00          770   
3  2487200875  20141209T000000   604000.0         4       3.00         1960   
4  1954400510  20150218T000000   510000.0         3       2.00         1680   
5  7237550310  20140512T000000  1225000.0         4       4.50         5420   
6  1321400060  20140627T000000   257500.0         3       2.25         1715   
7  2008000270  20150115T000000   291850.0         3       1.50         1060   
8  2414600126  20150415T000000   229500.0         3       1.00         1780   
9  3793500160  20150312T000000   323000.0         3       2.50         1890   

   sqft_lot  floors  waterfront  view  condition  grade  sqft_above  \
0      5650     1.0           0     0          3      7        1180   
1      7242     2.0           0     0          3      7        2170   
2     10000     1.0           0     0          3      6         770   
3      5000     1.0           0     0          5      7        1050   
4      8080     1.0           0     0          3      8        1680   
5    101930     1.0           0     0          3     11        3890   
6      6819     2.0           0     0          3      7        1715   
7      9711     1.0           0     0          3      7        1060   
8      7470     1.0           0     0          3      7        1050   
9      6560     2.0           0     0          3      7        1890   

   sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0              0      1955             0    98178  47.5112 -122.257   
1            400      1951          1991    98125  47.7210 -122.319   
2              0      1933             0    98028  47.7379 -122.233   
3            910      1965             0    98136  47.5208 -122.393   
4              0      1987             0    98074  47.6168 -122.045   
5           1530      2001             0    98053  47.6561 -122.005   
6              0      1995             0    98003  47.3097 -122.327   
7              0      1963             0    98198  47.4095 -122.315   
8            730      1960             0    98146  47.5123 -122.337   
9              0      2003             0    98038  47.3684 -122.031   

   sqft_living15  sqft_lot15  
0           1340        5650  
1           1690        7639  
2           2720        8062  
3           1360        5000  
4           1800        7503  
5           4760      101930  
6           2238        6819  
7           1650        9711  
8           1780        8113  
9           2390        7570

In [ ]:
# Print all the columns/features names (int)

# Q3.1 How many columns have `yr_` prefix?
# Q3.2 How many columns have `sqft_` prefix?
# Q3.3 How many columns assosiated with house location (except `zipcode`) are in the data?
# Q3.4 How many columns have `rooms` in their names?


In [ ]:
#Q3.2
import re
len(list(filter(lambda y: re.search('sqft_', y), df.columns)))

6

In [ ]:
#Q3.4
len(list(filter(lambda y: re.search('rooms', y), df.columns)))

2

In [ ]:
# Print data size (int)

# Q4.1 How many observations are in the data?
# Q4.2 How many features are in the data?

In [ ]:
#Answer for Q4.1
df.shape[0]

21613

In [ ]:
#Answer for Q4.2
df.shape[1]

21

# 2. Basic data exploration

Lets do some basics:
- `.count()` number of not NaN's in every column.
- Is there any missing values in the data?
- Count number of unique values in every column `.nunique()`.
- What does this tells you about the features, which are most likely categorical and which are most likely numerical?
- Use pandas `.describe()` to display basic statistic about the data.
- Use pandas `.value_counts()` to count number of unique values in a specific column.
- Use pandas `.min()`, `.max()`, `.mean()`, `.std()` to display specific statistics about the data.
- Use pandas `.dtypes` field to display data types in columns.

**Hint**
You could use `.sort_index()` or `.sort_values()` to sort the result of `.value_counts()`

In [ ]:
df.count()

id               21613
date             21613
price            21613
bedrooms         21613
bathrooms        21613
sqft_living      21613
sqft_lot         21613
floors           21613
waterfront       21613
view             21613
condition        21613
grade            21613
sqft_above       21613
sqft_basement    21613
yr_built         21613
yr_renovated     21613
zipcode          21613
lat              21613
long             21613
sqft_living15    21613
sqft_lot15       21613
dtype: int64

In [ ]:
df.isnull().sum()

id               0
date             0
price            0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
waterfront       0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
dtype: int64

In [ ]:
df.columns.nunique()

21

In [ ]:
df.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.161300e+04  2.161300e+04  21613.000000  21613.000000  21613.000000   
mean   4.580302e+09  5.400881e+05      3.370842      2.114757   2079.899736   
std    2.876566e+09  3.671272e+05      0.930062      0.770163    918.440897   
min    1.000102e+06  7.500000e+04      0.000000      0.000000    290.000000   
25%    2.123049e+09  3.219500e+05      3.000000      1.750000   1427.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    waterfront          view     condition  \
count  2.161300e+04  21613.000000  21613.000000  21613.000000  21613.000000   
mean   1.510697e+04      1.494309      0.007542      0.234303      3.409430   
std    4.142051e+04      0.539989      0.086517      0.766318      0.650743   
min    5.200000e+02      1.000000      0.000000      0.000000      1.000000   
25%    5.040000e+03      1.000000      0.000000      0.000000      3.000000   
50%    7.618000e+03      1.500000      0.000000      0.000000      3.000000   
75%    1.068800e+04      2.000000      0.000000      0.000000      4.000000   
max    1.651359e+06      3.500000      1.000000      4.000000      5.000000   

              grade    sqft_above  sqft_basement      yr_built  yr_renovated  \
count  21613.000000  21613.000000   21613.000000  21613.000000  21613.000000   
mean       7.656873   1788.390691     291.509045   1971.005136     84.402258   
std        1.175459    828.090978     442.575043     29.373411    401.679240   
min        1.000000    290.000000       0.000000   1900.000000      0.000000   
25%        7.000000   1190.000000       0.000000   1951.000000      0.000000   
50%        7.000000   1560.000000       0.000000   1975.000000      0.000000   
75%        8.000000   2210.000000     560.000000   1997.000000      0.000000   
max       13.000000   9410.000000    4820.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  
count  21613.000000  21613.000000  21613.000000   21613.000000   21613.000000  
mean   98077.939805     47.560053   -122.213896    1986.552492   12768.455652  
std       53.505026      0.138564      0.140828     685.391304   27304.179631  
min    98001.000000     47.155900   -122.519000     399.000000     651.000000  
25%    98033.000000     47.471000   -122.328000    1490.000000    5100.000000  
50%    98065.000000     47.571800   -122.230000    1840.000000    7620.000000  
75%    98118.000000     47.678000   -122.125000    2360.000000   10083.000000  
max    98199.000000     47.777600   -121.315000    6210.000000  871200.000000

In [ ]:
df.dtypes

id                 int64
date              object
price            float64
bedrooms           int64
bathrooms        float64
sqft_living        int64
sqft_lot           int64
floors           float64
waterfront         int64
view               int64
condition          int64
grade              int64
sqft_above         int64
sqft_basement      int64
yr_built           int64
yr_renovated       int64
zipcode            int64
lat              float64
long             float64
sqft_living15      int64
sqft_lot15         int64
dtype: object

In [ ]:
# Display number of not NaN's in every column (int)

# Q5.1 How many NA values are in the `floors` column?
# Q5.2 How many NA values are in the `grade` column?
# Q5.3 How many NA values are in the `bedrooms` column?
# Q5.4 How many NA values are in the `yr_built` column?
# Q5.5 How many explicit NA values are in the `yr_renovated` column?

In [ ]:
#Q5.1
df.floors.isna().sum()

0

In [ ]:
#Q5.5
df.yr_renovated.isna().sum()

0

In [ ]:
# Count number of unique values in every column (int)

# Q6.1 How many unique values are in the `bedrooms` column?
#df.bedrooms.nunique()
# Q6.2 How many unique values are in the `grade` column?
# Q6.3 How many unique values are in the `yr_renovated` column?
# Q6.4 How many unique values are in the `bathrooms` column?
#df.bathrooms.nunique()
# Q6.5 How many unique values are in the `long` column?


In [ ]:
#Q6.1
df.bedrooms.nunique()

13

In [ ]:
#Q6.4
df.bathrooms.nunique()

30

In [ ]:
# Count frequency of the values in different columns (list of ints in descending order)
# You could select a column using same syntax as for selecting a key from a dictionary: `data[colname]`
# Q7.1 For every unique `floors` value give its number of occurences.
# Q7.2 For every unique `condition` value give its number of occurences.
# Q7.3 For every unique `bedrooms` value give its number of occurences.
# Q7.4 For every unique `grade` value give its number of occurences.
# Q7.5 For every unique `view` value give its number of occurences.


In [ ]:
#Q7.3
df.bedrooms.value_counts().sort_values(ascending=False).values

array([9824, 6882, 2760, 1601,  272,  199,   38,   13,   13,    6,    3,
          1,    1])

In [ ]:
#Q7.5
df.view.value_counts().sort_values(ascending=False).values

array([19489,   963,   510,   332,   319])

In [ ]:
# Display basic data statistics using .describe()
df.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.161300e+04  2.161300e+04  21613.000000  21613.000000  21613.000000   
mean   4.580302e+09  5.400881e+05      3.370842      2.114757   2079.899736   
std    2.876566e+09  3.671272e+05      0.930062      0.770163    918.440897   
min    1.000102e+06  7.500000e+04      0.000000      0.000000    290.000000   
25%    2.123049e+09  3.219500e+05      3.000000      1.750000   1427.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    waterfront          view     condition  \
count  2.161300e+04  21613.000000  21613.000000  21613.000000  21613.000000   
mean   1.510697e+04      1.494309      0.007542      0.234303      3.409430   
std    4.142051e+04      0.539989      0.086517      0.766318      0.650743   
min    5.200000e+02      1.000000      0.000000      0.000000      1.000000   
25%    5.040000e+03      1.000000      0.000000      0.000000      3.000000   
50%    7.618000e+03      1.500000      0.000000      0.000000      3.000000   
75%    1.068800e+04      2.000000      0.000000      0.000000      4.000000   
max    1.651359e+06      3.500000      1.000000      4.000000      5.000000   

              grade    sqft_above  sqft_basement      yr_built  yr_renovated  \
count  21613.000000  21613.000000   21613.000000  21613.000000  21613.000000   
mean       7.656873   1788.390691     291.509045   1971.005136     84.402258   
std        1.175459    828.090978     442.575043     29.373411    401.679240   
min        1.000000    290.000000       0.000000   1900.000000      0.000000   
25%        7.000000   1190.000000       0.000000   1951.000000      0.000000   
50%        7.000000   1560.000000       0.000000   1975.000000      0.000000   
75%        8.000000   2210.000000     560.000000   1997.000000      0.000000   
max       13.000000   9410.000000    4820.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  
count  21613.000000  21613.000000  21613.000000   21613.000000   21613.000000  
mean   98077.939805     47.560053   -122.213896    1986.552492   12768.455652  
std       53.505026      0.138564      0.140828     685.391304   27304.179631  
min    98001.000000     47.155900   -122.519000     399.000000     651.000000  
25%    98033.000000     47.471000   -122.328000    1490.000000    5100.000000  
50%    98065.000000     47.571800   -122.230000    1840.000000    7620.000000  
75%    98118.000000     47.678000   -122.125000    2360.000000   10083.000000  
max    98199.000000     47.777600   -121.315000    6210.000000  871200.000000

In [ ]:
# Display some column statistics (list of floats, rounded up to 3 digits, e.g. 1.234)

# Q8.1 What are the max, min, mean and the std of the `floors` column?
# Q8.2 What are the max, min, mean and the std of the `bedrooms` column?
# Q8.3 What are the max, min, mean and the std of the `sqft_living` column?
# Q8.4 What are the max, min, mean and the std of the `pice` column?
# Q8.5 What are the max, min, mean and the std of the `long` column?

In [ ]:
#Q8.3
print("max sqft=","{:.3f}".format(df.sqft_living.max()))
print("min sqfts =","{:.3f}".format(df.sqft_living.min()))
print("mean  of sqfts=","{:.3f}".format(df.sqft_living.mean()))
print("std of sqft=","{:.3f}".format(df.sqft_living.std()))

max sqft= 13540.000
min sqfts = 290.000
mean  of sqfts= 2079.900
std of sqft= 918.441


In [ ]:
#Q8.5
print("max long=","{:.3f}".format(df.long.max()))
print("min long=","{:.3f}".format(df.long.min()))
print("mean long=","{:.3f}".format(df.long.mean()))
print("std long","{:.3f}".format(df.long.std()))


max long= -121.315
min long= -122.519
mean long= -122.214
std long 0.141


In [ ]:
# Display data types of all columns (int)

# Q9.1 How many columns have `object` data type?
# Q9.2 How many columns have `int64` data type?
# Q9.3 How many columns have `float64` data type?
# Display data types of all columns (list of str)
# Q9.4 What are the columns with dtype == `float64`?
# Q9.5 What are the columns with dtype == `int64`?

In [ ]:
#Q9.3
(df.dtypes == 'float64').sum()

5

In [ ]:
#Q9.4
df.columns[df.dtypes == 'float64']

Index(['price', 'bathrooms', 'floors', 'lat', 'long'], dtype='object')

In [ ]:
#optional code for question 9.4 :Answer
print("float data type:-",df.select_dtypes(include=['float64']).columns)

float data type:- Index(['price', 'bathrooms', 'floors', 'lat', 'long'], dtype='object')


# 3. Data selection

In pandas.DataFrame you could select

1. Row/s by position (integer number \[0 .. number of rows - 1\]) `.iloc` or by DataFrame.index `.loc`:
```
data.loc[0]
data.loc[5:10]
data.iloc[0]
data.iloc[5:10]
```
*Though, this is probably the worst way to manipulate rows.*

2. Columns by name
```
data[columname]
```
3. Row/s and columns
```
data.loc[10, columname]
data.iloc[10, columname]
```
4. Using boolean mask
```
mask = data[columname] > value
data[mask]
```
You could combine multiple conditions using `&` or `|` (and, or)

```
cond1 = data[columname1] > value1
cond2 = data[columname2] > value2
data[cond1 & cond2]
```
5. Using queries `.query()`:
```
value = 5
data.query("columname > value")
```
You could combine multiple conditions using `and`, `or`

```
data.query("(columname1 > value1) and (columname2 > value2)")
```
and others. See https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html for more examples.

*Remember to use different quotation marks " or ' for columnname inside a query.*


In [ ]:
# Select rows by position (int)

# Q10.1 How many bedrooms have a house on row 777?
# Q10.2 When was built a house on row 9999?
# Q10.3 How many floors have a house on row 1337?
# Q10.4 How many bathrooms have a house on row 314?
# Q10.5 What is the grade of a house on row 2718?


In [ ]:
#Q10.4
int(df['bathrooms'].iloc[313])

3

In [ ]:
#Q10.5
int(df['grade'].iloc[2717])

10

In [ ]:
## setting DataFrame index to be an `id` column, now .loc and .iloc will have different behavior
# Select rows by index (int)
df.index = df.id

# dropping `id` column, since now it is an index
df.drop('id', axis=1, inplace=True)

# sort data by index for clarity
df.sort_index(inplace=True)

# Q11.1 How many times the house with index 1000102 were sold?
# Q11.2 What is the price of the house with index 9842300095?
# Q11.3 When was built the house with index 104510440?
# Q11.4 What is the condition of a house with index 252000300?
# Q11.5 What is the living area (in square feets, feature 'sqft_living') of the house with index 1225069038?


In [ ]:
#Q11.1
df.loc[1000102].shape[0]

2

In [ ]:
#Q11.4
df.loc[252000300].condition

3

In [ ]:
# Using mask or .query syntax select rows/columns (int)

# Q12.1 How many houses were built during American Great Depression (1929–1939)? Including both start and end year.
# Q12.2 When was built the only house with basement area = 1024 sqft?
#df[df["sqft_basement"]==1024].yr_built
# Q12.3 How many houses are with the highest possible grade?
# Q12.4 When was built a house with maximal number of bedrooms?
# Q12.5 How many houses were sold for 256000 dollars?
#df[df["price"]==256000].value_counts()


In [ ]:
#Q12.2
df[df["sqft_basement"]==1024].yr_built

id
4136980090    2006
Name: yr_built, dtype: int64

In [ ]:
#Q12.5
(df["price"]==256000).sum()

8

In [ ]:
# Using mask or .query syntax select rows/columns (int)

# Q13.1 How many houses with the waterfront (=1) were built duroing Nixon's presidency (1969—1974)? Including both start and end year.
# Q13.2 How many houses, built before first human in space (<1961), have high condition (=5)?
# Q13.3 How many houses are with 6 bedrooms and less than 2000 sqft living area?
# Q13.4 What was the price of a house with 5 bathrooms, built in 1998 and graded with 10 score?
# Q13.5 How many floors has a house builtin 1999 with 5 bedrooms and 3400 sqft living area?

In [ ]:
#Q13.3
filtered_df = df[(df['bedrooms'] == 6) & (df['sqft_living'] < 2000)]
num_houses = len(filtered_df)
num_houses

20

In [ ]:
#Q13.4
df.query('bathrooms == 5 and yr_built == 1998 and grade == 10').price

5913    1180000.0
Name: price, dtype: float64

In [ ]:
# Using mask or .query syntax select rows/columns and compute simple statistics (float)

# Q14.1 What was the average (sold) price of a houses built in the year of Cuban Missile Crisis (1962)?
# Q14.2 What was the most expensive house built in the last decade of 20th centuary?
# Q14.3 What was the least expensive house built in the last decade of 20th centuary?
# Q14.4 What is the median number of bathrooms in houses with grade above 9 (10 and more)?
#df.query("grade>=9").bathrooms.median()
# Q14.5 What is the median grade of houses with most popular zipcode value?


In [ ]:
#Q14.3
last_decade_df = df[(df['yr_built'] >= 1990) & (df['yr_built'] <= 1999)]
least_expensive_house = last_decade_df.loc[last_decade_df['price'].idxmin()]
print("The least expensive house built in the last decade of the 20th century:")
print(least_expensive_house)


The least expensive house built in the last decade of the 20th century:
date             20141016T000000
price                   154000.0
bedrooms                       3
bathrooms                    3.0
sqft_living                 1530
sqft_lot                    9997
floors                       1.0
waterfront                     0
view                           0
condition                      3
grade                          6
sqft_above                  1020
sqft_basement                510
yr_built                    1992
yr_renovated                   0
zipcode                    98001
lat                      47.2861
long                    -122.252
sqft_living15               1410
sqft_lot15                  9997
Name: 3356403304, dtype: object


In [ ]:
#Q14.5
print(f'{df[df.zipcode == df.zipcode.value_counts().sort_values(ascending=False).index[0]].grade.median():.3f}')

7.000


# 4. Creating new columns


Creating new column of pandas.DataFrame is as easy as:
```
data['new_awesome_column'] = []
```
that's it. But such a column is relatively useless.
Typically, you would compute something new based on existing data and save it in a new column.
For example one might want to compute total area of the house as a sum of all `sqft_` columns, or
create a boolean column of whether the house has `grade` > 2 or anything else:

```
data['total_area'] = data[col1] + data[col2] + ...
data['high_value'] = data[col] > 5
```

Pandas also provides another powerfull tool: `.apply`, `.map()`, `.applymap()` methods (they are kinda the same, but not quite). https://stackoverflow.com/questions/19798153/difference-between-map-applymap-and-apply-methods-in-pandas .
They allow you to *apply* some function to every value in the column/s (row-wise) or row (column-wise) or cell (element-wise). For example, same computations of `total_area` and `high_value` using `.apply()`:
```
data['total_area'] = data[[col1, col2, col3]].apply(sum, axis=1)

```
you are not restricted to existent functions, `.apply()` accepts any function (including lambda functions):

```
data['total_area'] = data[[col1, col2, col3]].apply(lambda x: x[0]+x[1]+x[2], axis=1)
```
or ordinary python function (if this it should have complex behaviour):
```
def _sum(x):
    total = 0
    for elem in x:
        total += elem
    return total
    
data['total_area'] = data[[col1, col2, col3]].apply(_sum, axis=1)
```
Many pandas methods has `axis` parameter `axis=0` refers to rows, `axis=1` refers to columns.

*Warning. You should never use `for` loops to sum numerical elements from the container.*

---


In [ ]:
# Create `was_renovated` column. Bool column (0, 1) indicating whether the house was renovated.
df['was_renovated'] = (df.yr_renovated != 0).astype('int')

In [ ]:
# Create new columns using the old ones (new column in your DataFrame)

# Q15.1 Create a `sqft_tot_area` column (sum of all columns with `sqft_` prefix) using any method above
# Q15.2 Create a new column `sqm_tot_area` using `sqft_tot_area` and the fact that 1 foot = 0.3048 meters
# Q15.3 Create a new column `sqm_aver_floor_area` by dividing total area (in meters) by number of floors
# Q15.4 Create a new column `price_cat` by splitting a `price` into 5 ([1..5]) distinct intervals: 0 < x <=20%,
# 20% < x <= 40%, ... 80% < x <= 100% percentiles. You could use `.quantile()` to compute percentiles.
# Q15.5 Create a new bool column `high_class` it is True if the house has grade >= 9 and condition >= 4

In [ ]:
#Q15.1
sqft_tot_area = df[['sqft_living','sqft_lot','sqft_above' ,'sqft_basement','sqft_living15','sqft_lot15']].apply(sum, axis=1)
df['sqft_tot_area']=sqft_tot_area

In [ ]:
#Q15.2
df['sqm_tot_area'] = (df.apply(lambda x:x['sqft_tot_area']*0.3048,axis=1))

In [ ]:
#Q15.3
df['sqm_aver_floor_area'] = df.sqm_tot_area /(df.floors)

In [ ]:
#Q15.4
bins = df.price.quantile([0, 0.2, 0.4, 0.6, 0.8, 1]).values
labels = [i for i in range(1, 6)]

df['price_cat'] = pd.cut(df.price, bins=bins, labels=labels)
df['price_cat']

id
1000102       2
1000102       1
1200019       4
1200021       3
2800031       1
             ..
9842300095    2
9842300485    2
9842300540    2
9895000040    3
9900000190    1
Name: price_cat, Length: 21613, dtype: category
Categories (5, int64): [1 < 2 < 3 < 4 < 5]

In [ ]:
#Q15.5
df['high_class']=df.apply(lambda x:True if x['grade']>=9 and x['condition']>=4 else False,axis=1 )

In [ ]:
# Using mask or .query syntax select rows/columns (float)

# Q16.1 What is the average price of the house of the high_class(=True)?
# Q16.2 What is the average total_area (in meters) of the house from highest price category?
# Q16.3 What is the maximal number of floors amongst houses with the lowest price category?
# Q16.4 What is the most frequent zipcode amongst houses with the lowest price category?
# Q16.5 What is the minimal number of bathrooms in houses with high_class=True?


In [ ]:
#Q16.1
average_price = round(df[df['high_class'] == True]['price'].mean(),3)
print("The average price of houses with high class is:", average_price)

The average price of houses with high class is: 1146802.328


In [ ]:
#Q16.4
low_p_category = df['price_cat'].min()
low_p_houses = df[df['price_cat'] == low_p_category]
most_frequent_zipcode = float(low_p_houses['zipcode'].mode().iloc[0])
most_frequent_zipcode

98023.0

# 5. Basic date processing

You figure out that column `date` is to harsh for you, so you decided to convert it to a more plausible format:

- Use pandas method `to_datetime()` to convert the date to a good format.
- Exctract `year`, `month`, `day` and `weekday` from your new date column. Save them to separete columns.
- How many columns has your data now?
- Drop column `date`, remember to set `inplace` parameter to True.


**Hint** for datetime formatted date you could extract the `year` as follow:
```
data.date.dt.year
```

Very often date could be a ridiculously rich feature, sometimes it is holidays that matters, sometimes weekends, sometimes some special days like **black friday**.

Learn how to work with date in Python!

In [ ]:

# Create new columns based on `date` column
# Q17.1 Convert date to datetime format
# Q17.2 Extract and store `year`
# Q17.3 Extract and store `month`
# Q17.4 Extract and store `day`
# Q17.5 Extract and store `weekday`
# Q17.6 Create a new column `house_age_10` - the age of the house in full decades (e.g. 9 year old house - 0, 21 year old house - 2),
# using `yr_built` and 'year' columns

In [ ]:
#Q17.1
df.date = pd.to_datetime(df.date)

In [ ]:
#Q17.2
# Q17.2
df['year'] = df.date.dt.year

In [ ]:
#Q17.3
df['month'] =df.date.dt.month

In [ ]:
# Q17.4
df['day'] = df.date.dt.day

In [ ]:
# Q17.5
df['weekday'] = df.date.dt.weekday

In [ ]:
# Q17.6
df['house_age_10'] = (df.year - df.yr_built) // 10

In [ ]:
# Find some date related information from the data (int)

# Q18.1 What is the most popular selling weekday?
# Q18.2 What is the most popular selling month?
# Q18.3 What is the least popular selling weekday?
# Q18.4 What is the median age of the house (on a first available sold date)? (float)
# Q18.5 How many houses were sold on America's Independence Day (July, 4)?

In [ ]:
# Q18.4
int(df[df.date == df.date.min()].house_age_10.median())

3

In [ ]:
# Q18.5
df.loc[(df.day == 4) & (df.month == 7)].shape[0]

2

# 6. Groupby
From the documentation https://pandas.pydata.org/pandas-docs/stable/user_guide/groupby.html

By “group by” we are referring to a process involving one or more of the following steps:

    - Splitting the data into groups based on some criteria.
    - Applying a function to each group independently.
    - Combining the results into a data structure.
    
---
`.groupby()` is one of the most powerfull tool for feature engineering. Very often it is used to group object with the same categorical characteristics and compute some statistics (e.g. mean, max, etc.) of a their numerical characteric.

Instead of computing average area of houses with high grade you could compute average areas of the
houses for every grade in a single command:

```
data.groupby('grade')['sqm_tot_area'].mean()
```

You could also make multi-column groups:

```
data.groupby(['weekday','grade'])['price'].min()
```
next, you could compute multiple aggregation functions:
```
data.groupby(['weekday','grade'])['price'].agg([min, max])
```

instead of using built-in functions you could compute custom functions using apply:
```
import numpy as np
data.groupby(['condition','grade'])['bathrooms'].apply(lambda x: np.quantile(x, .5))
```

and the coolest thing now is that you can map the results of groupby back on your DataFrame!
```
gp = data.groupby(['condition'])['bathrooms'].median()
data['gp_feature'] = data['condition'].map(gp)
```
Now, if some house has `condition == 2`, its `gp_feature` will be equal to the median number of
bathrooms amongst all houses with `condition == 2`.

Read more examples in the documentation https://pandas.pydata.org/pandas-docs/stable/user_guide/groupby.html

In [ ]:
# Create some groupby features
# Q19.1 `price_by_class` groupby `high_class` and compute median `price`.
# Q19.2 `price_by_year` groupby `year` and compute median price.
# Q19.3 `price_by_weekday` groupby `weekday` and compute median price.
# Q19.4 `area_by_price` groupby `price_cat` and compute average `sqft_living`.
# Q19.5 'floors_by_age' groupby 'floors' and compute average age of a house using feature 'house_age_10'.

In [ ]:
#Q19.3
group_price_by_weekday = df.groupby(['weekday'])['price'].median()
df['price_by_weekday'] = df.year.map(group_price_by_weekday)

In [ ]:
# Q19.5
group_floors_by_age = df.groupby(['floors'])['house_age_10'].mean()
df['floors_by_age'] = df.floors.map(group_floors_by_age)


In [ ]:
# Create some other groupby features
# for this task check out this answer:
# https://stackoverflow.com/questions/47913343/how-to-groupby-and-map-by-two-columns-pandas-dataframe

# Q20.1 `n_houses_zipcode` groupby `zipcode` and count number of occurences of every unique zipcode
# Q20.2 `n_houses_yr_built` groupby `yr_built` and count number of houses built in each year
# Q20.3 `price_by_yr_month_`(median, std) groupby `year`, `month` and compute median and std `price`.
# Q20.4 `price_by_grade_age_`(median, std) groupby `grade`, `house_age` and compute median and std `price`.
# Q20.5 `living_by_cond_`(median, std) groupby `waterfront`, `view`, `condition` and compute median and std `sqft_living`.

In [ ]:
# Q20.3
group_price_by_yr_month_ = df.groupby(['year', 'month'])['price'].agg(['median', 'std'])

In [ ]:
group_price_by_yr_month_.columns = ['price_by_yr_month_median', 'price_by_yr_month_std']

In [ ]:
group_price_by_yr_month_ = group_price_by_yr_month_.reset_index()

In [ ]:
df = pd.merge(df, group_price_by_yr_month_, on=['year', 'month'], how='left')

In [ ]:
#Q20.5
group_by_living_cond_=df.groupby(['waterfront','view', 'condition'])['sqft_living'].agg(['median', 'std'])
group_by_living_cond_.columns =  ['living_by_cond_median','living_by_cond_std']
group_by_living_cond_=group_by_living_cond_.reset_index()
df=pd.merge(df,group_by_living_cond_,on=['waterfront','view', 'condition'],how='left')

# 7. Building a regression model

> You do not need to normalize data for tree models; for linear/knn models this step is essential.

> Remember, that not all of the features in the table are numeric, some of them might be viewed as categorical.

> You may create or drop **any** features you want, except for the features which use `price` (e.g. average price of a house with 5 bedrooms).

In [ ]:
# Q21 Drop all generated features which used price column, e.g. price_by_year, price_cat.
df = df.drop(columns=['price_cat','price_by_yr_month_median', 'price_by_yr_month_std'])


In [ ]:
threshold = 0.1
dominant_features = df.drop(columns=['date'])\
                 .columns[df.corr(numeric_only=True).price.sort_values(ascending=False) > threshold]

In [ ]:
dominant_features = dominant_features.drop('price')

In [ ]:
dominant_features

Index(['bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors',
       'waterfront', 'view', 'condition', 'grade', 'sqft_above',
       'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode', 'lat', 'long',
       'sqft_living15', 'sqft_lot15'],
      dtype='object')

In [ ]:
x = df[dominant_features].copy()
y = df[['price']].copy()

In [ ]:
# Q22 Split your data into train and test parts.
# How many records (rows) do you have in train and test tables? (list of int)?
# Use sklearn.model_selection.train_test_split with test_size=0.33 and random_state=42
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.33, random_state=42)
x_train.shape[0], x_test.shape[0]


(14480, 7133)

In [ ]:
# Create a predictive regression model of a house price.

# Q23.1 Use linear regression with l2 regularization (Ridge regression)
# Q23.2 Use decision tree regression
# Q23.3 Use k nearest neighbours regression

In [ ]:
#Q23.1
from sklearn.linear_model import Ridge
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline
from sklearn.metrics import r2_score

model_ridge = make_pipeline(StandardScaler(), Ridge())
model_ridge.fit(x_train, y_train)
print(f'R^2: {r2_score(model_ridge.predict(x_test), y_test):.3f}')

R^2: 0.617


In [ ]:
# Q23.3
from sklearn.neighbors import KNeighborsRegressor
model_knn = make_pipeline(StandardScaler(), KNeighborsRegressor())
model_knn.fit(x_train, y_train)
print(f'R^2: {r2_score(model_knn.predict(x_test), y_test):.3f}')

R^2: 0.732


In [ ]:
from sklearn.linear_model import Ridge

ridge = Ridge().fit(x_train, y_train)
print("Train Set R-square Val: {:.3f}".format(ridge.score(x_train, y_train)))
print("Test Set R-square Val: {:.3f}".format(ridge.score(x_test, y_test)))

ridge = Ridge(alpha=0.5).fit(x_train, y_train)
print("Train Set R-square Val: {:.3f}".format(ridge.score(x_train, y_train)))
print("Test Set R-square Val: {:.3f}".format(ridge.score(x_test, y_test)))

Train Set R-square Val: 0.695
Test Set R-square Val: 0.709
Train Set R-square Val: 0.695
Test Set R-square Val: 0.709


In [ ]:
# Use grid search to select optimal hyperparamters of your models.

# Q24.1 Alpha for a ridge regression
# Q24.2 Depth for the tree
# Q24.3 Number of neighbours for the knn

In [ ]:
#Q24.1
from sklearn.model_selection import GridSearchCV

param_grid = {'ridge__alpha': [j / 10 for j in range(11)]}
model_ridge = make_pipeline(StandardScaler(), Ridge())

gs_ridge = GridSearchCV(model_ridge, param_grid, scoring='r2', cv=5)
gs_ridge.fit(x_train, y_train)

print("Best params:", gs_ridge.best_params_)
print(f"Best score: {gs_ridge.best_score_:.3f}")

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_ridge.py:216: LinAlgWarning: Ill-conditioned matrix (rcond=2.75058e-17): result may not be accurate.
  return linalg.solve(A, Xy, assume_a="pos", overwrite_a=True).T
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_ridge.py:216: LinAlgWarning: Ill-conditioned matrix (rcond=1.26541e-17): result may not be accurate.
  return linalg.solve(A, Xy, assume_a="pos", overwrite_a=True).T


Best params: {'ridge__alpha': 1.0}
Best score: 0.692


In [ ]:
#Q24.3
param_grid = {'kneighborsregressor__n_neighbors': [2, 4, 6, 8, 10, 12, 14, 18, 20]}
model_knn = make_pipeline(StandardScaler(), KNeighborsRegressor())

gs_knn = GridSearchCV(model_knn, param_grid, scoring='r2', cv=5)
gs_knn.fit(x_train, y_train)

print("Best params:", gs_knn.best_params_)
print(f"Best score: {gs_knn.best_score_:.3f}")

Best params: {'kneighborsregressor__n_neighbors': 8}
Best score: 0.774


In [ ]:
# Compute train and test mean squared error for your best models (list of float).

# Q25.1 Train, test MSE using linear regression with l2 regularization
# Q25.2 Train, test MSE using decision tree regression
# Q25.3 Train, test MSE using k nearest neighbours regression


In [ ]:
#Q25.1
from sklearn.metrics import mean_squared_error
print(f'train MSE: {mean_squared_error(y_train, gs_ridge.predict(x_train)):.3f}')
print(f'test MSE:  {mean_squared_error(y_test, gs_ridge.predict(x_test)):.3f}')

train MSE: 42953004686.494
test MSE:  35710355754.152


In [ ]:
# Q25.3
print(f'train MSE: {mean_squared_error(y_train, gs_knn.predict(x_train)):.3f}')
print(f'test MSE:  {mean_squared_error(y_test, gs_knn.predict(x_test)):.3f}')

train MSE: 23965242758.833
test MSE:  23527843699.977


In [ ]:
# Compute train and test R^2 for your best models (list of float).
# Q26.1 Train, test R^2 using linear regression with l2 regularization
# Q26.2 Train, test R^2 using decision tree regression
# Q26.3 Train, test R^2 using k nearest neighbours regression


In [ ]:
# Q26.1
print(f'train R^2: {r2_score(y_train, gs_ridge.predict(x_train)):.3f}')
print(f'test R^2:  {r2_score(y_test, gs_ridge.predict(x_test)):.3f}')

train R^2: 0.695
test R^2:  0.709


In [ ]:
# Q26.3
print(f'train R^2: {r2_score(y_train, gs_knn.predict(x_train)):.3f}')
print(f'test R^2:  {r2_score(y_test, gs_knn.predict(x_test)):.3f}')

train R^2: 0.830
test R^2:  0.808


In [ ]:
# Q27 Which features have largest (by absolute value) weight in your linear model (top 5 features)? (list of str).
weights = gs_ridge.best_estimator_.named_steps['ridge'].coef_
feature_waights = sorted(list(zip(dominant_features, np.round(weights.ravel(), 3))), key=lambda x: abs(x[1]), reverse=True)
print([feature for feature, _ in feature_waights[:5]])

['grade', 'sqft_living', 'sqft_above', 'lat', 'yr_built']


# Make sure your .ipynb is linearly executable
# Kernel -> Restart & Run All -> No ERROR cells